content/en/docs/concepts/extend-kubernetes/api-extension/custom-resources.md

```diff
@@ -223,6 +224,7 @@ Aggregated APIs offer more advanced API features and customization of other feat
 | strategic-merge-patch | The new endpoints support PATCH with `Content-Type: application/strategic-merge-patch+json`. Useful for updating objects that may be modified both locally, and by the server. For more information, see ["Update API Objects in Place Using kubectl patch"](/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/) | No | Yes |
 | Protocol Buffers | The new resource supports clients that want to use Protocol Buffers | No | Yes |
 | OpenAPI Schema | Is there an OpenAPI (swagger) schema for the types that can be dynamically fetched from the server? Is the user protected from misspelling field names by ensuring only allowed fields are set? Are types enforced (in other words, don't put an `int` in a `string` field?) | Yes, based on the [OpenAPI v3.0 validation](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation) schema (GA in 1.16). | Yes |
+| Instance Name | Does this extension mechanism impose any constraints on the names of objects whose kind/resource is defined this way? | Yes, such an object's name must be a valid DNS subdomain name. | No |
```

Comment:
Just slightly more than a nit IMO:
```suggestion
| Instance Name | Does this extension mechanism impose any constraints on the names of resources whose kind/resource is defined this way? | Yes, such an object's name must be a valid DNS subdomain name. | No |
```

Aggregated APIs can serve things that aren't objects. Not sure about CRDs, maybe they can too (I can't think of an example though).

Reply:
A CRD can only define a kind of **object**.

Reply:
CRs are always objects with full CRUD. Non-persisted objects like SomethingReview you could implement in an aggregated apiserver are not possible with CRDs. So it's fine to talk about just objects for CRDs because all CRs are.

But even in the aggregated case, when you talk about "instance name" this by definition refers to `ObjectMeta` and that makes the resource an object.

Reply:
When an aggregated API server responds to a **get** or **list**, is this _always_ an object or collection? I'm thinking of MetricValue from custom.metrics.k8s.io/v1beta1 as a counterexample.

(We shouldn't assume that the aggregated API server uses our library or even what language it's written in, so long as it's conformant).

Reply:
If unsure, accept the suggestion.

Reply:
@sftim: please also beware that the term "aggregation" gets used for two **DIFFERENT** things, and I even confused myself when Slacking with @sttts. One of the two kinds of aggregation is when an APIService configures the kube-apiserver (or soon the generic control plane, yay!) to delegate requests to an external apiserver. The other kind of "aggregation" is when a CustomResouceDefinition object tells the kube-apiserver (or soon the generic control plane) how to store another kind of object. I will not bother to argue about the former (external server) here, but I think that is what you are referring to. CRDs are about the latter, and a CRD certainly can only define new kinds of _object_.

Reply:
I think there's a separate point about maybe coming up with a name for “the thing behind an APIService, where that APIService doesn't belong to the control plane“ (ie, an external API server). However, I don't want to merge something that could mislead readers into thinking that external API servers can only serve objects.

They actually can serve resources whether or not those resources have names (eg, MetricValueList collections).

Reply:
@sftim: _this_ PR has nothing to do with `APIService` objects and what they register. This PR is about what a `CustomResourceDefinition` object registers. They are two completely different things. (It is true that they both get called "aggregation", and I find this very aggrivating --- but that is not a battle to fight in the comments on this PR.)

Reply:
I think my suggestion is correct. If it's wrong, please explain why.

>  This PR has nothing to do with what can be delegated an external apiserver. 

This feedback is about a change to a row added to a table; that table compares using an APIService to using a CustomResourceDefinition. Are you sure that APIService isn't relevant? I'd like evidence about why, because I can't see it myself.

Reply:
> But even in the aggregated case, when you talk about "instance name" this by definition refers to ObjectMeta and that makes the resource an object.

@stts are you saying that for some resource fetched from a [conforming] extension API server using the **get** verb, if it can be fetched then it always has a `.metadata.name` (and is therefore an object)?

I know that typical implementations have `.metadata` and `.metadata.name` but I don't know that we guarantee it, or define conformance strictly enough that omitting it is formally an error. My best guess is that we don't make that guarantee, hence my suggestion.

Reply:
Oh, sorry @sftim, now I see why you are talking about the other form of aggregation.

I was hoping to avoid expanding the scope of the discussion. In fact, the suggested table edit is technically narrow: the first column holds a question, and in the suggested new row the question is only about _objects_. We _could_ leave it like that, the row only asking and answering a question about objects.

If you insist on expanding the discussion, I will answer the question about `APIService` and what it registers. Looking at the definition of an `APIService` object, you see that its `Spec` holds an API group that says what the `APIService` is about. Only objects belong to API groups, so an `APIService` can only set up delegation for some _objects_.

Reply:
Let's omit the change to this line for now. I still think it misleads, and if we omit this one change, we can move things forward.

In a separate PR, we can discuss whether aggregated APIs can only ever serve objects with names. Not here.

Reply:
Actually, I'm persuaded. It risks misleading but not very much. We can merge without clearing thing up.

/hold cancel